codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java

```diff
@@ -0,0 +1,203 @@
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+/**
+ * A {@link StreamByteDistributor} that ignores stream priority and uniformly allocates bytes to all
+ * streams. This class uses a minimum chunk size that will be allocated to each stream. While
+ * fewer streams may be written to in each call to {@link #distribute(int, Writer)}, doing this
+ * should improve the goodput on each written stream.
+ */
+public final class UniformStreamByteDistributor implements StreamByteDistributor {
+    static final int DEFAULT_MIN_ALLOCATION_CHUNK = 1024;
+
+    private final Http2Connection.PropertyKey stateKey;
+    private final Deque<State> queue = new ArrayDeque<State>();
+    private final Deque<State> emptyFrameQueue = new ArrayDeque<State>();
+
+    /**
+     * The minimum number of bytes that we will attempt to allocate to a stream. This is to
+     * help improve goodput on a per-stream basis.
+     */
+    private int minAllocationChunk = DEFAULT_MIN_ALLOCATION_CHUNK;
+    private long totalStreamableBytes;
+
+    public UniformStreamByteDistributor(Http2Connection connection) {
+        checkNotNull(connection, "connection");
+
+        // Add a state for the connection.
+        stateKey = connection.newKey();
+        Http2Stream connectionStream = connection.connectionStream();
+        connectionStream.setProperty(stateKey, new State(connectionStream));
+
+        // Register for notification of new streams.
+        connection.addListener(new Http2ConnectionAdapter() {
+            @Override
+            public void onStreamAdded(Http2Stream stream) {
+                stream.setProperty(stateKey, new State(stream));
+            }
+
+            @Override
+            public void onStreamClosed(Http2Stream stream) {
+                state(stream).close();
+            }
+        });
+    }
+
+    /**
+     * Sets the minimum allocation chunk that will be allocated to each stream. Defaults to 1KiB.
+     *
+     * @param minAllocationChunk the minimum number of bytes that will be allocated to each stream.
+     * Must be > 0.
+     */
+    public void minAllocationChunk(int minAllocationChunk) {
+        if (minAllocationChunk <= 0) {
+            throw new IllegalArgumentException("minAllocationChunk must be > 0");
+        }
+        this.minAllocationChunk = minAllocationChunk;
+    }
+
+    @Override
+    public void updateStreamableBytes(StreamState streamState) {
+        State state = state(streamState.stream());
+        state.updateStreamableBytes(streamState.streamableBytes(), streamState.hasFrame());
+    }
+
+    @Override
+    public boolean distribute(int maxBytes, Writer writer) throws Http2Exception {
+        checkNotNull(writer, "writer");
+
+        // First, write out any empty frames.
+        while (!emptyFrameQueue.isEmpty()) {
+            State state = emptyFrameQueue.remove();
```

Comment:
@nmittler - With the current 2 queue approach is seems like we only get the benefit of not having to iterate through the entire queue is if for a given stream the `streamableBytes == 0`. This condition will be false the first time a "non-empty" stream has been added since the last write operation. What if a stream has the following frames sizes: `1024, 0` before a write operation?  In this case is it possible we will service the first 1024 byte frame, and not service the 0 byte frame (assume the two frames cannot be merged, e.g. data and trailing headers)?

I wonder if we could simplify, and have a lower memory footprint, by just using 1 queue and either always iterate through it (what is the perf implications), or maybe tracking a `boolean queueHasEmptyFrame` to be able to short circuit the full iteration in some situations. WDYT?